### PR TITLE
Added support for reading arrays of arbitrary JSON values

### PR DESCRIPTION
**Pull request by lilleyse**
## Description

I was getting some warnings when parsing [`keySet`](https://github.com/CesiumGS/3d-tiles/blob/ad47ad1f7e91b3d4deb9edfc42cb1293b5334306/extensions/3DTILES_content_conditional/schema/tileset.3DTILES_content_conditional.schema.json#L19-L26) in [`3DTILES_content_conditional`](https://github.com/CesiumGS/3d-tiles/pull/834):

```
[2026-07-07 14:41:18.714] [warning] CesiumNativeWriteTilesetError: errors A string value is not allowed in the object array and has been replaced with a default value.

  While parsing: .extensionsRequired.dimensions[1].keySet[0]
  From byte offset: 28192, A string value is not allowed in the object array and has been replaced with a default value.
  While parsing: .extensionsRequired.dimensions[2].keySet[0]
  From byte offset: 28285, A boolean value is not allowed in the object array and has been replaced with a default value.
  While parsing: .extensionsRequired.dimensions[3].keySet[0]
  From byte offset: 28354
```

`keySet` can be an array of strings, numbers, or boolean, so its generated type is

```c++
  CesiumJsonReader::ArrayJsonHandler<
      CesiumUtility::JsonValue,
      CesiumJsonReader::JsonObjectJsonHandler>
      _keySet;
```

This doesn't work with `ArrayJsonHandler` because it expects its elements to be objects (hence the warnings above). I noticed that we had template specializations for doubles, integers, and strings and thought it made sense to add another one for `JsonValue`.

## Author checklist

- [x] I have submitted a [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) (only needed once).
- [x] I have done a full self-review of my code.
- [x] I have updated `CHANGES.md` with a short summary of my change (for user-facing changes).
- [ ] I have added or updated unit tests to ensure consistent code coverage as necessary.
- [x] I have updated the documentation as necessary.

We don't have any unit tests in `CesiumJsonReader` :/

## Reviewer checklist

Thank you for taking the time to review this PR. By approving a PR you are taking as much responsibility for these changes as the author.

As you review, please go through the checklist below:

- [ ] Review and run all parts of the test plan on this branch and verify it matches expectations.
    - [ ] If the issue is a bug please make sure you can reproduce the bug in the main branch and then checkout this branch to make sure it actually solved the issue.
- [ ] Review the code and make sure you do not have any remaining questions or concerns. You should understand the code change and the chosen approach. If you are not confident or have doubts about the code, please do not hesitate to ask questions.
    - [ ] Code should follow the [C++ Style Guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html)
- [ ] Review the unit tests and make sure there are no missing tests or edge cases.
- [ ] Review documentation changes and updates to `CHANGES.md` to make sure they accurately cover the work in this PR.
- [ ] Verify that the [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) has been submitted, if needed.

### CODE REVIEW

**Comment by kring** [comment]
Looks good to me. I'll let the core team merge it though.

File: CesiumJsonReader/include/CesiumJsonReader/ArrayJsonHandler.h [hunk 1]
```diff
@@ -144,6 +146,172 @@ class CESIUMJSONREADER_API ArrayJsonHandler : public JsonHandler {
   std::unique_ptr<THandler> _objectHandler;
 };
 
+/**
+ * @brief Special case of \ref ArrayJsonHandler for handling arrays of arbitrary
+ * JSON values.
+ */
+template <>
+class CESIUMJSONREADER_API
```
Comment (j9liu) [inline CesiumJsonReader/include/CesiumJsonReader/ArrayJsonHandler.h:154]: Can we get a unit test for this? I know they don't exist for this library, but we really should add some :')

**Review by j9liu — COMMENTED** [review]
Thanks @lilleyse ! This looks good to me from a glance, though I trust @kring 's vetting more 😄 

Just one comment about unit tests!

**Comment by kring** [comment]
@copilot please add unit tests for the various specializations of `ArrayJsonHandler`. The tests should verify that 1) each handler is able to read JSON of the expected type, and 2) the handlers generate sensible warnings for unexpected JSON, and 3) unexpected JSON elements are skipped, so encountering them does not break the ability to read later valid elements.

File: CesiumJsonReader/test/TestArrayJsonHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+using namespace CesiumJsonReader;
+using namespace CesiumUtility;
+
+namespace {
+
+template <typename THandler>
+auto readArray(std::string_view json) {
+  THandler handler;
+  auto bytes = std::span<const std::byte>(
```
Comment (kring) [inline CesiumJsonReader/test/TestArrayJsonHandler.cpp]: Avoid `auto`. Use `std::span<const std::byte> bytes(...)` instead.

File: CesiumJsonReader/test/TestArrayJsonHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+class TestObjectJsonHandler : public ObjectJsonHandler {
+public:
+  using ValueType = TestObject;
+
+  TestObjectJsonHandler() noexcept = default;
+
+  void reset(IJsonHandler* pParent, TestObject* pObject) {
+    JsonHandler::reset(pParent);
+    _pObject = pObject;
```
Comment (kring) [inline CesiumJsonReader/test/TestArrayJsonHandler.cpp]:
```suggestion
    this->_pObject = pObject;
```

File: CesiumJsonReader/test/TestArrayJsonHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    CHECK((*result.value)[0].isInt64());
+    CHECK((*result.value)[0].getInt64() == 1);
+    CHECK((*result.value)[1].isString());
+    CHECK((*result.value)[1].getString() == "hello");
+    CHECK((*result.value)[2].isBool());
+    CHECK((*result.value)[2].getBool() == true);
+    CHECK((*result.value)[3].isNull());
+    CHECK((*result.value)[4].isDouble());
+    CHECK((*result.value)[4].getDouble() == 2.5);
+    CHECK((*result.value)[5].isObject());
```
Comment (kring) [inline CesiumJsonReader/test/TestArrayJsonHandler.cpp:283]: Check the content of the object is correct, too.

File: CesiumJsonReader/test/TestArrayJsonHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+    CHECK(result.warnings.empty());
+    REQUIRE(result.value->size() == 2);
+    CHECK((*result.value)[0].isObject());
+    CHECK((*result.value)[1].isObject());
+    const auto* pA = (*result.value)[0].getValuePtrForKey("a");
+    REQUIRE(pA != nullptr);
+    CHECK(pA->getInt64() == 1);
+  }
+
+  SUBCASE("mixed types - booleans strings and numbers (original bug case)") {
```
Comment (kring) [inline CesiumJsonReader/test/TestArrayJsonHandler.cpp]: Remove the "original bug case" comment. Outside the context of this PR it does not make sense.

**Review by kring — CHANGES_REQUESTED** [review]

**Comment by kring** [comment]
@copilot address my review comments, run clang-format on the new test file, and fix the failing test in the recent CI runs. You can probably avoid building and running the code yourself, but if you need to, use the same steps that CI uses on your platform by looking at the `.github./workflows/build.yml` file.

**Comment by kring** [comment]
Copilot added some decent tests and found a bug (sorry I missed this in my review). It then fixed it in a ridiculous way, so I fixed it in a better way. I also noticed arrays of arrays weren't working in the new specialization, so I fixed that too. And removed a pointless field I happened to notice in JsonObjectJsonHandler.

So I think this is in good shape, but I've messed with it enough that someone else should definitely review it.

**Review by kring — APPROVED** [review]

**Review by j9liu — APPROVED** [review]
Thanks @lilleyse and @kring !
